src/Microsoft.AspNetCore.Authentication/AuthenticationHandler.cs

```diff
@@ -118,8 +118,17 @@ protected AuthenticationHandler(IOptionsMonitor<TOptions> options, ILoggerFactor
         protected string BuildRedirectUri(string targetPath)
             => Request.Scheme + "://" + Request.Host + OriginalPathBase + targetPath;
 
+        protected virtual string ResolveTarget(string scheme)
+            => scheme ?? Options.ForwardDefaultSelector?.Invoke(Context) ?? Options.ForwardDefault;
```

Comment:
Updated added tests and checks to prevent reentering the methods, we don't support concurrency in our handlers today right?  Each request should have its own handler by default...